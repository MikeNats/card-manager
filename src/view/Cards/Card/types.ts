import { CardModel } from '../../../domain/card/entity'

export type CardProps= {
  cardData: CardModel
  toggleModal:Function
  edit: boolean
}