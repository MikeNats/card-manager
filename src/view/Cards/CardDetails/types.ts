import { CardModel } from '../../../domain/card/entity'

export type CardDetailsProps = { 
    cardData: CardModel,
}