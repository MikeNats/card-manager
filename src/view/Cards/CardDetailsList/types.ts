
import { CardModel } from '../../../domain/card/entity'

export type CardDetailsListProps={
    cardCollection: CardModel[]
}