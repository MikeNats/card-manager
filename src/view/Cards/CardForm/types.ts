import { CardModel } from '../../../domain/card/entity'
import {ReactText} from 'react'
export type CardFormProps = { 
  operation: OPERATION,
  cardData: CardModel | CardModelInput,
  newInstance: boolean
}

export enum OPERATION{
  EDIT= 'EDIT',
  CREATE= 'CREATE'
}

export interface CardModelInput{
  cardNumber: string
  id: string;
  type: string;
  name: string;
  expiration: string; 
  cvc: ReactText;
}

