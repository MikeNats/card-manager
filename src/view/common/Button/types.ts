export type ButtonProps = {
  id?:string
  className?: string,
  children?: string,
  onClickHandler?: Function
  disabled?: boolean
} 