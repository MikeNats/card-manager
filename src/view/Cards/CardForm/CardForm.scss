.comp-card-form{
    margin-top: 15px;
    .comp-card{
        margin-bottom: 35px;
    }
    fieldset{
        position: relative;
    }
    label{
        font-weight: bold;
        font-size: 14px;
        color: #000;
        margin-bottom: 14px;
    }
    input{
        font-size: 14px;
        color: #000;
        width:100%;
        border: none;
        border-bottom: 1px solid $grey-20;
        padding-bottom: 10px;
        outline: none;
    }
    ::placeholder { 
        color: grey-20;
      } 
    input {
        background-position: 100% 0%;
        background-repeat: no-repeat;
    }
    input.error{
        background-image: url('/form-error.svg');
    }
    .valid input{
        background-image: url('/form-success.svg');
        border-color: $green-30;
        color: $green-30;  
    }
    input.error, .errorMessage{
        color: $red-30;
        border-color: $red-30;

    }
    .errorMessage{
        margin:0;
        font-size: 14px;
        position: absolute;
        bottom: -12px;    
    }
    .errorMessage.hide{
        display: none;
    }
    button:disabled {
        background-color: $grey-20;
    }
    button{
        margin: 0 auto 48px auto;
    }
    button:focus{
        outline: none;
    }
    #delete-card-button{
        background: none;
        color: $grey-20;
        padding:0;
        &:hover{
            color: #000;
        }
    }
}