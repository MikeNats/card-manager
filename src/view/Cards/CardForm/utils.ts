import { OPERATION, CardModelInput } from './types'
import {createCard} from '../../../domain/card/cardService' 
import {createNewCard, updateCrad} from '../../../service/Cards'


export const createUpdateCard = (operation: OPERATION, cardData: CardModelInput ):Promise<any> => 
   operation === OPERATION.EDIT ? 
      updateCrad({
        id: cardData.id,
        type: cardData.type,
        name:cardData.name, 
        cardNumber: +`${cardData.cardNumber}`.replace(/\s/g, ""),
        expiration: cardData.expiration,
        cvc: +`${cardData.cvc}`})
   :
      createNewCard(createCard({
        id: '', 
        type: '',
        name:cardData.name, 
        cardNumber: +`${cardData.cardNumber}`.replace(/\s/g, ""),
        expiration: cardData.expiration,
        cvc: +`${cardData.cvc}`}))



