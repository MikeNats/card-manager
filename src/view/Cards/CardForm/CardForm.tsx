import React, {useEffect, useState} from 'react';
import { isValidCardName, isValidCardNumber, isValidDate, isValidCvc } from '../../../domain/card/cardService'
import Input from '../../common/Input/Input'
import { throttle } from 'lodash'
import {CardFormProps, OPERATION} from './types' 
import Button from '../../common/Button/Button'
import {  createUpdateCard } from './utils'
import { deleteCrad } from '../../../service/Cards'
import { addSpaceEveryFourChars, addSlashEveryTwoChars } from '../../../domain/card/cardService'
  

const CardForm = function({operation, cardData, newInstance}: CardFormProps) {
  const [name, setName] = useState(cardData.name || '')
  const [cardNumber, setCardNumber] = useState(addSpaceEveryFourChars(cardData.cardNumber || ''))
  const [expiration, setExpiration] = useState(cardData.expiration || '')
  const [cvc, setCvc] = useState(cardData.cvc || '')  

  const [invalidname, setInvalidName] = useState(false)
  const [invalidCardNumber, setInvalidCardNumber] = useState(false)
  const [invalidexpiration, setInvalidexpiration] = useState(false)
  const [invalidCvc, setInvalidCvc] = useState(false)
  const [asyncError, setAsyncError] = useState(false)
  const [validName, seValidName] = useState(false)
  const [validCardNumber, setValidCardNumber] = useState(false)
  const [validexpiration, setValidexpiration] = useState(false)
  const [validCvc, setValidCvc] = useState(false)
   
  useEffect(() => {
     setAsyncError(false)
    if(operation === OPERATION.CREATE) {
      setName('') 
      setCardNumber('')
      setExpiration('')
      setCvc('')
    }

    seValidName(false)
    setValidCardNumber(false) 
    setValidexpiration(false)
    setValidCvc(false)
  }, [newInstance, operation])

  const submitForm = (e: React.FormEvent<HTMLFormElement>) =>{
    e.preventDefault();
    validateForm();
    if (!invalidname &&
      !invalidCardNumber &&
      !invalidexpiration &&
      !invalidCvc) {
        createUpdateCard(operation,
          {id: cardData.id, 
          type: cardData.type,
          name, cardNumber, 
          expiration,  
          cvc}).catch(e => {
            setAsyncError(true)})
      }
  }
  const deleteCard = () => 
    deleteCrad(cardData.id)
      .catch(e => setAsyncError(true))

  const onnameChange = throttle(name => 
    setName(name), 200)

  const onCardNumberChange = throttle(inputCardNumber => 
    setCardNumber(addSpaceEveryFourChars(inputCardNumber)), 200)

  const onexpirationChange = throttle(inputexpiration => 
    setExpiration(addSlashEveryTwoChars(inputexpiration)), 200) 

  const onCvcChange = throttle(inputCvc => 
    setCvc(inputCvc), 200)

  const validateForm = () => {
    nameValidation()
    cardNumberValidation()
    expirationValidation()
    cvcValidation()
  }

  const nameValidation = () => {
    if (!isValidCardName(name)) {
      setInvalidName(true)
      seValidName(false)
    } else {
      setInvalidName(false)
      seValidName(true)
    }
  }

  const cardNumberValidation = () => {
    if (!isValidCardNumber(cardNumber)) {
      setInvalidCardNumber(true)
      setValidCardNumber(false)
    } else {
      setInvalidCardNumber(false)
      setValidCardNumber(true)
    }
  }
      

  const expirationValidation = () => {
    if(!isValidDate(expiration)) {
      setInvalidexpiration(true) 
      setValidexpiration(false)
    } else {
      setInvalidexpiration(false)
      setValidexpiration(true)
    }
  }

  const cvcValidation = () =>{
   if(!isValidCvc(cvc)) {
    setInvalidCvc(true)
    setValidCvc(false)
   } else {
    setInvalidCvc(false)
    setValidCvc(true)
   }
  }
   
  return (<section className="comp-card-form">
    { asyncError ? <h3>Oops somthing went wrong...</h3> :
      <>
      <form onSubmit={submitForm}>
          <Input 
              title="Name in card"
              placeholder="Jhon Cabruci"
              type="text"
              id="card-name"
              name="card-name"
              errorMessage="Please fill your name"
              className={validName ? 'valid' : ''}
              onBlur={nameValidation}
              value={name}
              error={invalidname}
              onChangeHandler={onnameChange}/>
          <Input 
            title="Card number"
            type="text"
            placeholder="5532 1234 5545 8014"
            id="card-number"
            name="card-number"
            errorMessage="Please enter a valid card number"
            className={validCardNumber ? 'valid' : ''}
            onBlur={cardNumberValidation}
            maxLength={19}  
            value={cardNumber}  
            error={invalidCardNumber}
            onChangeHandler={onCardNumberChange}/>
          <Input
            title="Expirecy date"
            placeholder="08/21"
            type="text"
            id="expirecy-date"
            name="expirecy-date"
            maxLength={5} 
            errorMessage="Please enter a valid expirecy date"
            className={validexpiration ? 'valid' : ''}
            onBlur={expirationValidation}
            value={expiration}
            error={invalidexpiration}
            onChangeHandler={onexpirationChange}/>
          <Input
            title="CVC (Security code)"
            placeholder="009"
            type="text" 
            id="cvc"
            name="cvc"
            errorMessage="Please enter a valid security code"
            className={validCvc ? 'valid' : ''}
            onBlur={cvcValidation}
            value={`${cvc}`}
            maxLength={3}  
            error={invalidCvc}
            onChangeHandler={onCvcChange}/>

          <Button 
            id="create-update-card-button" 
            disabled= {!!(
              !name &&
              !cardNumber &&
              (!expiration && expiration.split("").length !==5 )  &&
              (!cvc && `${cvc}`.split("").length !==3 ))
            }>Confirm</Button> 
      </form>
     <Button 
        id="delete-card-button"
        disabled= {!cardData.id}
        className={operation === OPERATION.EDIT ? '' : 'hide'} 
        onClickHandler={deleteCard}>Delete Card</Button></>
      }
    </section>
  ) 
} 

export default React.memo(CardForm)