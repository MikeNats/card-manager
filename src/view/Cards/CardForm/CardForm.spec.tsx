import React from 'react'
import CardForm from './CardForm'
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { CardModel } from '../../../domain/card/entity'
import {OPERATION} from './types'
import * as services from'../../../service/Cards'
jest.mock('../../../service/Cards');

Enzyme.configure({ adapter: new Adapter() })
const mockCard: CardModel ={
  id: "1231234234",
  type: "mastercard",
  name: "Jhon Done",
  cardNumber: "5500000000000004",
  expiration: "10/24",
  cvc:"432"
};

async function tick() {
  return await new Promise(resolve =>
    setTimeout(resolve, 1000))
} 

describe("CardForm", () => {

  it('renders as expected', () => {
    const component = shallow(<CardForm
      cardData={mockCard}
      operation={OPERATION.EDIT}
  />)

    expect(component).toMatchSnapshot();
  })
  describe("Validations", () => {
    const validateDeleteButtonStatus = (component: any, assertValue: boolean, operation: OPERATION) => 
      operation === OPERATION.EDIT ? expect(component.find("#delete-card-button").props().disabled).toEqual(assertValue) : 
      expect(component.find("#delete-card-button").props().className).toEqual('hide');
    [OPERATION.CREATE, OPERATION.EDIT].forEach((operation: OPERATION) => {
      it(`should not allow to submit if card name is empty - ${operation}`, () => {
        const component = shallow(<CardForm
          newInstance={true}
          cardData={Object.assign({},mockCard, {name: ""})}
          operation={operation}
      />)  
        validateDeleteButtonStatus(component, false, operation)
        expect(component.find("#create-update-card-button").props().disabled).toEqual(false)
      })

     it(`should not allow to submit if card name is three not letters followed by space last name three letters - ${operation}`,  () => {
      ["a", "aa", "aaa ", "aaa a", "aaa aa"].forEach(senarion => {
        const component = shallow(<CardForm
          newInstance={true}
          cardData={Object.assign({}, mockCard, {name:senarion})}
          operation={operation} 
        />) 
        component.find("#card-name").simulate('blur', {preventDefault: ()=>{}})

        tick().then(() =>{
          expect(component.find("#card-name").props().error).toEqual(true)
        })

      })  
    })
      
      it(`should not allow to submit if card number is empty - ${operation}`, () => {
        const component = shallow(<CardForm
          newInstance={true}
          cardData={Object.assign({},mockCard, {cardNumber: ""})} 
          operation={operation}
      />) 

      tick().then(() =>{
        validateDeleteButtonStatus(component, false, operation)
        expect(component.find("#create-update-card-button").props().disabled).toEqual(true)      })
      })
  
      it(`should not allow to submit if card expiration is empty - ${operation}`, () => {
        const component = shallow(<CardForm
          newInstance={true}
          cardData={Object.assign({},mockCard, {expiration: ""})} 
          operation={operation}
      />)  
      tick().then(() =>{
        expect(component.find("#create-update-card-button").props().disabled).toEqual(true)
          validateDeleteButtonStatus(component, false, operation)})
      })

      it(`should not allow to submit if card expiration is not mm/yy - ${operation}`, () => {
        const component = shallow(<CardForm
          newInstance={true}
          cardData={Object.assign({},mockCard, {expiration: "80/20"})} 
          operation={operation}
      />)  
      tick().then(() =>{
        component.find("#expirecy-date").simulate('blur', {preventDefault: ()=>{}})
        expect(component.find("#expirecy-date").props().error).toEqual(true)})
      })
  
      it(`should not allow to submit if card cvc is empty - ${operation}`, () => {
        const component = shallow(<CardForm 
          newInstance={true}
          cardData={Object.assign({},mockCard, {cvc: ""})} 
          operation={operation}
      />)   

      tick().then(() =>{
        expect(component.find("#create-update-card-button").props().disabled).toEqual(true)
          validateDeleteButtonStatus(component, false, operation)})
      })

      it(`should not allow to submit if card cvc can not parsed to int - ${operation}`, () => {
        services.createNewCard.mockImplementation(() => Promise.resolve());
        services.updateCrad.mockImplementation(() => Promise.resolve());
        const component = shallow(<CardForm 
          cardData={Object.assign({},mockCard, {cvc: "erwe"})}  
          operation={operation}
      />)  
       
        tick().then(() =>{
          component.find("form").first().simulate('submit', {preventDefault: ()=>{}})
          expect(component.find("#cvc").props().error).toEqual(true)})
        })
     it(`should not allow to submit if card data is not valid - ${operation}`, () => {
        const component = shallow(<CardForm
          cardData={Object.assign({},mockCard, {cardNumber:"wdeadc45s6c"})}
          operation={operation} 
      />)  
       
      tick().then(() =>{
          component.find("#card-number").simulate('blur', {preventDefault: ()=>{}})
          expect(component.find("#card-number").props().error).toEqual(true)})
      })

      it(`should submit if card data is valid - ${operation}`, () => {
        const component = shallow(<CardForm
          cardData={mockCard} 
          operation={operation} 
      />)  
        expect(component.find("#create-update-card-button").props().disabled).toEqual(false)
          validateDeleteButtonStatus(component, false, operation)
      })
    })
    describe("Submition", () => {
      afterEach(() => {    
        jest.clearAllMocks();
      });
      describe("Create", () => {
        it('should invoke createNewCard', async () => {
          services.createNewCard.mockImplementation(() => Promise.resolve());
          const component = shallow(<CardForm
            cardData={mockCard} 
            operation={OPERATION.CREATE} 
        />)  
 
          component.find("form").first().simulate('submit', {preventDefault: ()=>{}})
          expect(services.createNewCard).toBeCalled()
        }) 
      })
      describe("Update", () => { 
        it('should invoke updateCrad', async () => {
          services.updateCrad.mockImplementation(() => Promise.resolve());
          const component = shallow(<CardForm
            cardData={mockCard} 
            operation={OPERATION.EDIT} 
        />)  

          component.find("form").first().simulate('submit', {preventDefault: ()=>{}})
          expect(services.updateCrad).toBeCalled()
        })
      }) 
      describe("Delete", () => {
        it('should invoke createNewCard', async () => {
          services.deleteCrad.mockImplementation(() => Promise.resolve());
          const component = mount(<CardForm
            cardData={mockCard} 
            operation={OPERATION.EDIT}  
        />)   
          component.find("#delete-card-button").find('button').simulate('click')
          expect(services.deleteCrad).toBeCalled()
        })
      })
    })

  })
})
