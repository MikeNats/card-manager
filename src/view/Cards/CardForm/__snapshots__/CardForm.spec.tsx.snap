// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardForm renders as expected 1`] = `ShallowWrapper {}`;
